### PR TITLE
chore(web): updates build product references, extracts CI test-build scripts 📜

### PR DESCRIPTION
**Pull request by jahorton**
This 📜 PR set is for some long-needed maintenance:  the Web project's files are being reorganized into a format more consistent with the other Keyman projects.  There's no new functionality; the end-goal is for everything to be stable and working as before (at the end of the 📜 PR chain).

Continues from #7556.

At this point in time, note that only "Test - 14.0" test config steps have been extracted.  There's more work to be done to also handle nightly / beta / stable test configs properly.

Now that all the `i`'s have been dotted and `t`'s crossed for properly reconnecting everything to our CI configurations (including on the CI server side!)...

### Other Notes

Re: the last would-be test defined below...

https://github.com/keymanapp/keyman/blob/c91875bbb19000cd2b54ad8637854cb680a04db8/web/src/tools/testing/recorder/recorder_KeyboardScripts.js#L1-L11

We'll need an extra configuration here to handle this page when hosted from the CI server; build artifacts aren't hosted from the site's root, but from a significantly-deep subfolder instead.  The stub-hosting part was "simple enough", but... the second half - loading from those stubs - may take some extra work.

image

## User Testing

As this 📜 PR chain has shuffled a lot of Web files around, the main goal here is to ensure that everything is linked properly in its new location.

**TEST_ANDROID_LAUNCH**:  Verify that the Keyman for Android app launches successfully and displays the keyboard.
- Any Android device, emulated or real, will suffice.

**TEST_IOS_LAUNCH**:  Verify that the Keyman for iPhone and iPad app launches successfully and displays the keyboard.
- Any iOS device, simulated or real, will suffice.

**TEST_WEB_PREDICTIONS**:  Using Chrome **remote device emulation**...
- Navigate through the following links:
    - KeymanWeb Test Home (as linked [here](https://github.com/keymanapp/keyman/pull/7831#issuecomment-1328653844))
    - "View Keymanweb website-oriented manual test pages"
    - "Prediction UI testing"
- Click the "Type here" field.
- Verify that the OSK appears and that the predictive-text banner offers suggestions.

**TEST_ANDROID_HARNESS**:  Using Chrome **remote device emulation**...
- Navigate through the following links:
    - KeymanWeb Test Home (as linked [here](https://github.com/keymanapp/keyman/pull/7831#issuecomment-1328653844))
    - "View manual test pages for embedded-mode Keymanweb"
    - "Android test harness"
- Verify that an OSK appears at the bottom of the black box.
- Using the OSK, type `a` and verify that an `a` is emitted in the text area above the black box.

**TEST_WEB_SAMPLE_TOOLBAR**:  Using a standard desktop browser (and not emulating a touch device)...
- Navigate through the following links:
    - KeymanWeb Test Home (as linked [here](https://github.com/keymanapp/keyman/pull/7831#issuecomment-1328653844))
    - "View Keymanweb use samples"
    - "Full multilingual (compiled) Keymanweb"
- Verify that a "toolbar" of sorts appears on the page.
   - Wait a couple of seconds before proceeding; when additional options appear on the toolbar, proceed.
- Click the toolbar's "Languages" button and verify that a world map appears and that a list of languages is available for selection.

_**Known issue with the following test: needs extra configuration work.  May be delayed as a separate issue.**_

Skip this; I've ensured that the `test-bot` will too.

_SKIP_TEST_WEB_RECORDER_:  Using a standard desktop browser...
- Navigate through the following links:
    - KeymanWeb Test Home (as linked [here](https://github.com/keymanapp/keyman/pull/7831#issuecomment-1328653844))
    - "Use Keymanweb development tools."
    - "Test-sequence recorder"
- Verify that no alerts occur when loading the page.
- Scroll down to the "Input Recording" section, then click within the textbox immediately below "Record input here".
- A small menu should appear to the right-hand side of the textbox - use it to select "Simple Deadkeys (English)".
- Type `a` with the OSK, then `s` with your computer's hardware keyboard.
    - New data should show up in the big area to the right-hand side - new data for every keystroke.
    - You should see a `"type": "osk"` entry before any `"type": "key"` entry.

### CODE REVIEW

File: web/src/test/manual/embed/android-harness/build.sh [hunk 1]
```diff
@@ -21,6 +21,10 @@ cp "$KEYMAN_ROOT/web/build/app/embed/debug/osk/kmwosk.css" "$THIS_DIR/host/kmwos
 cp "$KEYMAN_ROOT/web/build/app/embed/debug/osk/globe-hint.css" "$THIS_DIR/host/globe-hint.css"
 cp "$KEYMAN_ROOT/web/build/app/embed/debug/osk/keymanweb-osk.ttf" "$THIS_DIR/host/keymanweb-osk.ttf"
 
+# We shouldn't rely on a prior Android build for these files!  We can retrieve them separately.
+cp "$KEYMAN_ROOT/common/web/sentry-manager/build/index.js" "$THIS_DIR/host/keyman-sentry.js"
+cp "$KEYMAN_ROOT/node_modules/es6-shim/es6-shim.min.js" "$THIS_DIR/host/es6-shim.min.js"
```
Comment (jahorton) [inline web/src/test/manual/embed/android-harness/build.sh:26]: While these two files aren't particularly necessary for direct functionality for the Android-harness test page, the errors that occur when trying to load and use them - `keyman-sentry.js`, in particular - seem to cascade into other errors that bring the whole house of cards down.

Worth noting:  up until this PR chain, we never actually 'published' this testing page on the testing index.  Looking up an old PR's artifacts, this didn't work before, either.  That said... I think a little "elbow grease" can overcome this and get this page operational for future user tests.

**Comment by bharanidharanj** [comment]
* **TEST_ANDROID_LAUNCH (PASSED):** Tested with the attached PR build (17.0.13-alpha-test-7831) in API 31 / Android 12.0 emulator and I was able to see the keyboard after launching Keyman app.

**Comment by bharanidharanj** [comment]
* **TEST_IOS_LAUNCH (PASSED):** Tested with the attached PR build in iOS 15.0 / iPhone 12 Pro Max and iPad Air Tablet Simulators and I was able to see the keyboard after launching Keyman app. 


..iPhone 12 Pro Max 

   

..iPad Air Tablet

**Comment by bharanidharanj** [comment]
* **TEST_WEB_PREDICTIONS (PASSED):** Tested in the given link with chrome device emulator and I was able to see the OSK that appears with the Predictive texts.

**Comment by bharanidharanj** [comment]
* **TEST_ANDROID_HARNESS (PASSED):** Tested in the given link in chrome device emulator and here is my observation: 1. Verified that the OSK appears at the bottom of the black box. 2. Typed letter 'a' appears in the text area which appears above the black box.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_WEB_SAMPLE_TOOLBAR (PASSED):** Tested this in the given web link and verified that a world map appears and that a list of languages was available for selection after clicking the toolbar's "Languages" button. Seems to be working as expected.

File: web/src/tools/testing/recorder/recorder_KeyboardScripts.js [hunk 2]
```diff
@@ -1,16 +1,32 @@
...
+    // `/resources` should be rooted on that.
+
+    let testSiteRootSuffixMatcher = /Keymanweb_TestPullRequests\/\d{1,10}:id/;
+    let match = testSiteRootSuffixMatcher.exec(location.href);
+    if(match && match.length > 0) {
+      let totalLength = match.index + match[0].length;
+      RESOURCE_PATH_PREFIX = location.href.substring(0, totalLength);
+    } else {
+      console.error("Failed to properly detect recorder-resource path for TC build artifact");
+    }
```
Comment (mcdurdin) [inline web/src/tools/testing/recorder/recorder_KeyboardScripts.js:26]: 😍

**Review by mcdurdin — APPROVED** [review]
LGTM

File: web/ci.sh [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+if builder_start_action validate-size; then
+  # Performs the web build product size check as reported on Web test PRs.
+  FLAGS=
+
+  if ! builder_has_option --debug; then
+    FLAGS=--write-status
+  fi
+
+  # Do not use --report when in --debug mode; it sets an error code that is VERY difficult to
+  # catch with our set -e & `trap` setup.
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/ci.sh:82]: ?

https://github.com/keymanapp/keyman/blob/f5acaf6e08d656698a639946cd9e8ce8652987ee/web/src/tools/building/check-build-size.sh#L35-L43

I almost set the default for `FLAGS` to `FLAGS=--report`.  This turned out to be a "bad idea," hence the comment.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.41-alpha](https://keyman.com/downloads/releases/alpha/17.0.41)

**Pull request merged**